cli/fathom_web/extract.py

```diff
@@ -0,0 +1,144 @@
+    subresources_directory = file.parent / 'resources' / f'{file.stem}_resources'
+    subresources_directory.mkdir(parents=True, exist_ok=True)
+
+    offset = 0
+    new_html = ''
+
+    filename_counter = 0
+    # A cache for elements that are repeated (e.g. icons)
+    saved_strings = {}
+
+    # Remove any existing `<base>` tag
+    html = BASE_TAG_PATTERN.sub('', html)
+
+    # Add `'self'` to the Content Security Policy so we can load our extracted resources
+    html = OLD_CSP.sub(NEW_CSP, html)
+
+    base64_data_matches = BASE64_DATA_PATTERN.finditer(html)
+    for match in base64_data_matches:
+        # Add the content before the base64 data
+        new_html += html[offset:match.start(1)]
+
+        base64_string = match.group('string')
+
+        # Check to see if we have already encountered this base64 string
+        file_path = saved_strings.get(base64_string, None)
+        if file_path is None:
+            mime_type = match.group('mime')
+            filename_counter += 1
+            filename = generate_filename(mime_type, str(filename_counter))
+            binary_data = decode_base64_string_to_binary(base64_string)
+            file_path = subresources_directory / filename
+            save_binary_data(binary_data, file_path)
+            saved_strings[base64_string] = file_path
+
+        # "Replace" the old base64 data with the relative path to the newly created file
+        new_html += file_path.relative_to(file.parent).as_posix()
+
+        # Move our offset to the end of the old base64 data
+        offset = match.end(1)
+
+    # Add the remainder of the content
+    new_html += html[offset:]
+
+    with file.open('w', encoding='utf-8') as fp:
+        fp.write(new_html)
+
+
+def decode_base64_string_to_binary(base64_string: str) -> bytes:
+    """
+    Decodes a string containing base64 data generated by Freeze-dry into bytes and returns it.
+
+    Freeze-dry generates strings of the form: 'data:[some_mime_type];base64,[the_actual_base64_string]'
+    """
+    return base64.b64decode(base64_string)
+
+
+def generate_filename(mime_type: str, filename: str) -> str:
+    """
+    Creates a filename to use for saving the base64 data with the appropriate file extension.
+
+    We can't necessarily get an appropriate filename from the html the base64 data comes from, so we don't
+    even try. Instead we just an incrementing counter to ensure unique filenames.
+
+    The appropriate extension comes from mapping the MIME type contained in the base64 data to an extension.
+    """
+    # `mimetypes` gets some extensions wrong (e.g. image/jpeg -> .jpe) and doesn't
+    # work for some MIME types that freeze-dry gives so use our own mapping first
+    try:
+        extension = MIME_TYPE_TO_FILE_EXTENSION[mime_type]
+    except KeyError:
+        extension = mimetypes.guess_extension(mime_type, strict=True)
+    return f'{filename}{extension}'
+
+
+def save_binary_data(binary_data: bytes, filepath: pathlib.Path):
```

Comment:
Do this and decode_base64_string_to_binary exist as separate functions just to give you a place to hang docstrings?

Reply:
`decode_base64_string_to_binary()` changed throughout developing this and wasn't originally a one-liner. I can inline that. As for `save_binary_data()`, I tend to pull out pieces like this that add indentation levels so the main code (in this instance the body of the `for` loop starting on line 80) reads easily at a high level. But, I do have the saving of the file on lines 106 and 107 right there in the code, so I suppose I should be fine inlining this as well.

Reply:
I don't have a strong opinion. I've written the same function before. In my case, I did it because I was writing a whole bunch of literals to various files, and it saved a line for each one.

Reply:
Upon inlining, it looks clear enough to me, so that's what will happen in the next commit.